src/net_processing.cpp

```diff
@@ -2438,9 +2441,27 @@ void ProcessMessage(
         return;
     }
 
-    if (msg_type == NetMsgType::ADDR) {
+    if (msg_type == NetMsgType::ADDR || msg_type == NetMsgType::ADDRv2) {
+        const auto version_orig = vRecv.GetVersion();
+        if (msg_type == NetMsgType::ADDRv2) {
+            vRecv.SetVersion(version_orig | ADDRv2_FORMAT);
+        }
         std::vector<CAddress> vAddr;
-        vRecv >> vAddr;
+
+        try {
+            vRecv >> vAddr;
+        } catch (const std::exception& e) {
+            vRecv.SetVersion(version_orig);
+
+            LOCK(cs_main);
+            Misbehaving(pfrom.GetId(), 20,
+                strprintf(
+                    "Supplied us with a message that contains an unparsable address: %s",
+                    e.what()));
+            return;
+        }
```

Comment:
This part probably deserves some extra scrutiny, I know that it's generally better to avoid exceptions in C++, but when I first wrote this I didn't see another option here for bubbling up an error and handling it.

Reply:
Yes, I also couldn't find a better way to cancel the parsing. I think that is a limitation of our ser/unser framework - its not possible to cancel the parsing of multiple entries in the middle (e.g. when parsing a `std::vector` of `CNetAddr`) without using an exception.

Another way, but that's not good, would be to not throw an exception from `CNetAddr` unserialize methods, consume the bogus data from the stream, mark the object as invalid, continue parsing the remaining `CNetAddr` objects and later remove the invalid objects. But that is not good because we do not want to continue parsing the data once we see that it is clearly malformed/malicious.

Reply:
@sipa Could you advise on whether there's another way or if the current way is alright?